source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
Oh, I'm fine with -1 right up until it causes a RELEASE_ASSERT in the connection constructor. Let's chat offline because this is totally a case where unvalidated config could easily cause runtime crashes and I don't think we want that.